ipfs-9-to-10/migration/config_conv.go

```diff
@@ -0,0 +1,189 @@
+	confExpi := confMap["Experimental"].(map[string]interface{})
+	if confExpi == nil {
+		return false, false
+	}
+	enabledi, ok := confExpi["QUIC"]
+	enabled := enabledi.(bool)
+
+	delete(confExpi, "QUIC")
+
+	return enabled, ok
+}
+
+// Convert Bootstrap addresses to/from QUIC
+func convertBootstrap(confMap map[string]interface{}, conv convArray) {
+	bootstrapi, _ := confMap["Bootstrap"].([]interface{})
+	if bootstrapi == nil {
+		log.Log("Bootstrap field missing or of the wrong type")
+		return
+	}
+	bootstrap := make([]string, len(bootstrapi))
+	for i := range bootstrapi {
+		bootstrap[i] = bootstrapi[i].(string)
+	}
+	confMap["Bootstrap"] = conv(bootstrap)
+}
+
+// Convert Addresses.Swarm to/from QUIC
+func convertSwarm(confMap map[string]interface{}, conv convArray) {
+	addressesi, _ := confMap["Addresses"].(map[string]interface{})
+	if addressesi == nil {
+		log.Log("Addresses field missing or of the wrong type")
+		return
+	}
+
+	swarmi, _ := addressesi["Swarm"].([]interface{})
+	if swarmi == nil {
+		log.Log("Addresses.Swarm field missing or of the wrong type")
+		return
+	}
+
+	swarm := make([]string, len(swarmi))
+	for i := range swarmi {
+		swarm[i] = swarmi[i].(string)
+	}
+	addressesi["Swarm"] = conv(swarm)
+}
+
+// Add QUIC Bootstrap address
+func ver9to10Bootstrap(bootstrap []string) []string {
+	hasOld := false
+	hasNew := false
+	res := make([]string, 0, len(bootstrap)+1)
+	for _, addr := range bootstrap {
+		res = append(res, addr)
+		if addr == ip4BootstrapAddr {
+			hasOld = true
+		} else if addr == quicBootstrapAddr {
+			hasNew = true
+		}
+	}
+
+	// If the config has the old IP v4 bootstrapper, add the new QUIC
+	// bootstrapper
+	if hasOld && !hasNew {
+		res = append(res, quicBootstrapAddr)
+	}
+
+	return res
+}
+
+func ver10to9Bootstrap(bootstrap []string) []string {
+	// No need to remove the QUIC bootstrapper, just leave things as they are
+	return bootstrap
+}
+
+var tcpRegexp = regexp.MustCompile(`/tcp/([0-9]+)`)
+
+// For each TCP listener, add a QUIC listener
+func ver9to10Swarm(swarm []string) []string {
+	res := make([]string, 0, len(swarm)*2)
+	for _, addr := range swarm {
+		res = append(res, addr)
+
+		// If the old configuration already has a quic address in it, assume
+		// the user has already set up their swarm addresses for quic and leave
+		// things as they are
+		if strings.Contains(addr, "/udp/quic") {
+			return swarm
+		}
+	}
+
+	// For each tcp address, add a corresponding quic address
+	for _, addr := range swarm {
+		if tcpRegexp.MatchString(addr) {
+			res = append(res, tcpRegexp.ReplaceAllString(addr, `/udp/$1/quic`))
+		}
+	}
+
+	return res
+}
+
+var quicRegexp = regexp.MustCompile(`/udp/[0-9]+/quic`)
+
+// Remove QUIC listeners
+func ver10to9Swarm(swarm []string) []string {
+	// Remove quic addresses
+	res := make([]string, 0, len(swarm))
+	for _, addr := range swarm {
+		if !quicRegexp.MatchString(addr) {
+			res = append(res, addr)
+		}
+	}
+
+	return res
+}
```

Comment:
This will remove QUIC listeners that the user had setup. I guess if we want to not remove that we should keep a marker indicating if we changed anything on ver9to10Swarm or not.